build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js

```diff
@@ -0,0 +1,102 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Global typedef map typedefName: typedef comment
+const TYPEDEFS = new Map();
+
+const buildTypedefs = (t, path) => {
+  const typedefs = [];
+  for (const [typedefName, typedefComment] of TYPEDEFS) {
+    const ast = buildVarDeclAndComment(t, path, typedefName, typedefComment);
+    typedefs.push(ast);
+  }
+  return typedefs;
+};
+
+const buildVarDeclAndComment = (t, path, name, comment) => {
+  const decl = t.variableDeclaration('let', [
+    t.variableDeclarator(t.identifier(name)),
+  ]);
+  // Add the typedef annotation.
+  t.addComment(decl, 'leading', comment);
+  return decl;
+};
+
+module.exports = function(babel) {
+  const {types: t} = babel;
+  // `shouldWriteToFile` should only be true in the production pipeline.
+  let shouldWriteToFile;
+  // `shouldEmitTypedefs` is used for testing purposes only.
+  let shouldEmitTypedefs;
+  return {
+    pre() {
+      TYPEDEFS.clear();
+      const {writeToFile = false, emitTypedefs = false} = this.opts;
+      shouldWriteToFile = writeToFile;
+      shouldEmitTypedefs = emitTypedefs;
+    },
+    visitor: {
+      Program: {
+        exit(path) {
+          if (shouldWriteToFile) {
+            // Stub
+            console.log('write to temp directory');
+          }
+          if (shouldEmitTypedefs) {
+            // Preserve the leading LICENSE comment.
+            path.addComment('leading', path.parent.comments[0].value);
+
+            const typedefs = buildTypedefs(t, path);
+            path.replaceWith(t.program(typedefs));
+            path.skip();
+          }
+        },
+      },
+      VariableDeclaration(path) {
+        const {node} = path;
+
+        if (!node.leadingComments) {
+          return;
+        }
+
+        const typedefComment = node.leadingComments.find(comment => {
+          return (
+            comment.type === 'CommentBlock' && /@typedef/.test(comment.value)
+          );
+        });
+
+        if (!typedefComment) {
+          return;
+        }
+
+        // We can assume theres only 1 variable declaration  when a typedef
+        // annotation is found. This is because Closure Compiler does not allow
+        // declaration of multiple variables with a shared type information.
+        const typedefName = node.declarations[0].id.name;
+
+        const typedefLocation = TYPEDEFS.get(typedefName);
+        if (!typedefLocation) {
+          TYPEDEFS.set(typedefName, typedefComment.value);
+        }
+
+        // We can't easily remove comment nodes so we just empty the string out.
+        typedefComment.value = '';
```

Comment:
No need to do this if we're removing the path.

Reply:
doesn't seem to remove the leading comments if i don't empty it out:

see https://astexplorer.net/#/gist/d501eb1193748a09482f7cdd69e1870e/ba9a293cedd23f9c8b64bd16d114e0cee2d503cf